### PR TITLE
Lock channels sending a message

### PR DESCRIPTION
**Pull request by Half-Shot**
This is a proposal for an alternative way to avoid message echoback.

### CODE REVIEW

File: src/bot.ts [hunk 1]
```diff
@@ -136,6 +138,34 @@ export class DiscordBot {
...
 
+    public lockChannel(channel: Discord.GuildChannel) {
+        const key = `${channel.guild.id}|${channel.id}`;
+        if (this.channelLocks[key]) {
+            return;
+        }
+        let i: NodeJS.Timeout;
+        const p = new Promise((resolve) => {
+            i = setInterval(resolve, this.config.limits.discordSendDelay);
+            this.channelLocks[key] = {i, p};
```
Comment (Sorunome) [inline src/bot.ts]: shouldn't this be outside the promise?

File: src/bot.ts [hunk 2]
```diff
@@ -136,6 +138,34 @@ export class DiscordBot {
         return this.provisioner;
     }
 
+    public lockChannel(channel: Discord.GuildChannel) {
+        const key = `${channel.guild.id}|${channel.id}`;
```
Comment (Sorunome) [inline src/bot.ts]: `channel.id` is already a unique key accross all guilds

File: src/bot.ts [hunk 3]
```diff
@@ -200,7 +229,7 @@ export class DiscordBot {
 
         client.on("messageDelete", async (msg: Discord.Message) => {
             try {
-                await Util.DelayedPromise(this.config.limits.discordSendDelay);
+                await this.waitUnlock(msg.channel as Discord.TextChannel);
```
Comment (Sorunome) [inline src/bot.ts]: if we just use `channel.id` as key we don't need to cast here, either (which might lead to errors if people DM the discord bot etc.)

File: src/bot.ts (hunk 1 above)
Comment (Half-Shot) [reply src/bot.ts]: Nope, because we need to set the interval pointer *after* it's been set, and `setInterval` needs the resolve function exposed.

**Pull request merged**